### PR TITLE
Compare metrics-based stats for every envoy spans, using a heatmap

### PR DESCRIPTION
**Pull request by jotak**
It uses new endpoint `api/stats/metrics`

Fixes https://github.com/kiali/kiali/issues/3293

Required backend PR: https://github.com/kiali/kiali/pull/3347

### CODE REVIEW

**Comment by jotak** [comment]
In traces details page / spans list:

Capture d’écran de 2020-10-30 11-49-06

File: src/components/JaegerIntegration/JaegerResults/StatsComparison.tsx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import * as API from 'services/Api';
+import { Color, HeatMap } from 'components/HeatMap/HeatMap';
+import { MetricsStatsQuery, Target } from 'types/MetricsOptions';
+import { EnvoySpanInfo } from '../JaegerHelper';
+import { SpanTableItem } from './SpanTableItem';
+import { MetricsStats } from 'types/Metrics';
+import { PfColors } from 'components/Pf/PfColors';
+import { Button, ButtonVariant, Tooltip } from '@patternfly/react-core';
+
+// TODO / follow-up: these variables are to be removed from here and managed from the toolbar
```
Comment (jotak) [inline src/components/JaegerIntegration/JaegerResults/StatsComparison.tsx:14]: About this TODO: I created a task: https://github.com/kiali/kiali/issues/3379

**Comment by lucasponce** [comment]
Adding @lwrigh.

@jotak conceptually the heatmap in the Span looks good, I like it, but I think it's taking too much space in the row and personally I see it weird.

I suggest to add a topic in the UX agenda for next Tuesday and see if Laura can help.

to @lwrigh, feel free also to add questions and/or commens even to catch up with the context.

**Comment by jotak** [comment]
@lucasponce yes, I wanted to see it more horizontal than vertical as well, so I did several tries, like switching the axis (but it's weird to have durations on Y) or changing default stats to show more durations and less percentiles (but IMO this is less interesting).

Also note that, if we do the follow-up task that I'm planning ( https://github.com/kiali/kiali/issues/3379 ), percentiles and durations to show could be changed, which means that the shape of the heatmap will change as well.

**Comment by jotak** [comment]
(probably we can remove the p75 line by default, it will still be vertical, but not as much as now)

**Comment by lwrigh** [comment]
PatternFly 4 does not currently have a heat map component but should be a component that's available in the future. PatternFly 3 has a heat map component if you want to use that instead so it's more consistent with the rest of the UI: https://www.patternfly.org/v3/pattern-library/data-visualization/heat-map/index.html 

If you don't want to change the current heat map component that is being used I would suggest to update the colors to be more consistent with PatternFly at least. PatternFly colors for charts: https://www.patternfly.org/v4/guidelines/colors-for-charts

We can talk further about the different options in the next UX meeting.

**Comment by jotak** [comment]
Thanks @lwrigh

The problem about the colors is that because the nature of the heatmap, these are gradient computed mathematically, so they won't strictly match pre-predefined colors, except perhaps at the boundaries. Having a heatmap built from a limited set of colors (like in PF3) removes all the tones that I think are important for this kind of dataviz. If we take a look at grafana for instance, heatmaps show gradiant colors: https://www.google.com/search?q=grafana+heatmap

We can discuss it in the next meeting

**Comment by jotak** [comment]
I'm still thinking about it .. trying to see if we could map pf4 color families to a scale that goes, more or less linearly, from plain red (worst) to plain green (best). If there is a decent number of tones, it worth it to try. I'm annoyed with the "green" family, I think it only have 3 usable tones out of the 5 proposed.

**Comment by lwrigh** [comment]
> I'm still thinking about it .. trying to see if we could map pf4 color families to a scale that goes, more or less linearly, from plain red (worst) to plain green (best). If there is a decent number of tones, it worth it to try. I'm annoyed with the "green" family, I think it only have 3 usable tones out of the 5 proposed.

I would also take a closer look at how accessible some of the color combinations are too. The contrast of the text against the square background color doesn't look like there's enough contrast for some of them.

**Comment by hhovsepy** [comment]
@jotak can these comparison maps be expandable per row?
Screenshot from 2020-11-02 11-29-53

**Comment by jotak** [comment]
@hhovsepy not sure what you mean, you'd like them to not be displayed by default, but only on demand?

**Comment by hhovsepy** [comment]
I mean hide these column content by default, and instead show 'show more info' link, and expand the table only when clicking 'show more info' link. Just my idea from UX perspective, like we do for labels in lists.

**Comment by lucasponce** [comment]
> I mean hide these column content by default, and instead show 'show more info' link, and expand the table only when clicking 'show more info' link. Just my idea from UX perspective, like we do for labels in lists.

I'm also inclined to think that we are accumulating a lot of good info per row, so perhaps a table/expandable pattern can suit on it http://patternfly-react.surge.sh/components/table#expandable

About the heatmap, I'm undecided I see several drawbacks:
- It uses a lot of space.
- Also, it delegates to the user to but the absolute value of the span inside of the heatmap to compare.

So, my overall suggestion:
- Use a table/expandable pattern.
- Try to see if some custom heatmap or minigraph suits, where it's clear there is the span request time compared with the stats.

Some UI suggestions, but here I delegate on @lwrigh comments, I personally see very clean a similar pattern from Github:
image

I'd also included in a different contrast where the span is located inside, and I'd use more tooltip to get details, instead to put too much explicit legends.

Goal is to get the attention of the user in the span that it may require further investigation.

**Comment by jotak** [comment]
Here is another try, with a couple of changes:

- Although it's still uses gradient / color interpolation, I'm using a bunch of patternfly colors as "key frames", so the result looks more "patternflyish".
- Remove p75 to make it smaller
- Removed cell borders

Capture d’écran de 2020-11-04 11-12-38

**Comment by mtho11** [comment]
This is quite cool! Perhaps in the future it could even be made more interactive, such that the time period (duration) is adjustable via a slider. This way you could see it change in time by just moving the slider. Similar to our existing **Graph Replay** technology. IMO it would make a really cool demo feature.

**Comment by jotak** [comment]
Here's a new screenshot:

Capture d’écran de 2020-11-09 13-47-45

Changes are:
- Swapped axis (to make it horizontal by default)
- margin & border radius on cells
- reversed interval order (10m to 6h)
- increased text contrast

@lwrigh I'm not sure if the patternfly table component can have sticky headers as we discussed, I couldn't find anything like that ( https://www.patternfly.org/2020.04/documentation/react/components/table )

**Comment by jotak** [comment]
> This is quite cool! Perhaps in the future it could even be made more interactive, such that the time period (duration) is adjustable via a slider. This way you could see it change in time by just moving the slider. Similar to our existing **Graph Replay** technology. IMO it would make a really cool demo feature.

Thanks, I've mentioned it in this follow-up task, because it wouldn't be for that PR: https://github.com/kiali/kiali/issues/3379

**Comment by jotak** [comment]
Another follow-up tasks entered to the backlog: https://github.com/kiali/kiali/issues/3420 => for expandable mode

**Comment by jotak** [comment]
BTW, personally I prefer the version with plain squares & no margin, wdyt?

**Comment by lucasponce** [comment]
I've fetched locally and started to review.

A first comment, it seems that it needs a wide resolution, in 1500px x 1200px I got this displacement in the titles of the heatmap:

image

**Comment by lucasponce** [comment]
Note, probably for a new issue, but a "Killer" feature would be to put a heatmap in a floating layer when clicking a trace, like:
image

**Comment by jotak** [comment]
> Note, probably for a new issue, but a "Killer" feature would be to put a heatmap in a floating layer when clicking a trace

Good idea. But since the comparisons are done per span, not per trace, we need to aggregate all compared spans stats (like: take the worst of each stat? the average?). Anyway it's worth thinking about it !

**Review by lucasponce — APPROVED** [review]
LGTM
If @hhovsepy can give a final test we can consolidate.

I have a naming comment in the backend code, but easy, I think.

**Comment by lucasponce** [comment]
> Good idea. But since the comparisons are done per span, not per trace, we need to aggregate all compared spans stats (like: take the worst of each stat? the average?). Anyway it's worth thinking about it !

Perhaps it's simpler for a first step, look at the chart:

image

As a user I have an average of "25ms", which is seems slow compared with others, but the heatmap would tell me if that "25ms" is very bad or something okey based on the context.

Obviously analytics can go deep as we can, but I was refering using the current data (trace avg time used to place it in the chart, and heatmap calculated from stats).

Anyway, my thinking has always same topic: how to locate the span I'm interested on.

**Comment by jotak** [comment]
> A first comment, it seems that it needs a wide resolution, in 1500px x 1200px I got this displacement in the titles of the heatmap:

Fixed, adding `nowrap` to css

**Review by hhovsepy — APPROVED** [review]
Verified in all traces tabs:
Screenshot from 2020-11-12 13-03-42

**Comment by jotak** [comment]
Thanks!

**Pull request merged**